### ISSUE TITLE
Component and View Colocation

### ISSUE THREAD

**Issue by BlakeWilliams** [ViewComponent/view_component#618]
Currently, components live in a dedicated `app/components` directory. This has worked well so far, but we'd like to explore a more unified approach.

This isn't set in stone, but the goals would be to:

* Allow components to be defined in `app/views`
* Enable `render "users/show", user: current_user` to render a component instead of explicitly referring to the component class. e.g. This would be equivalent to `render Users::Show.new(user: current_user)`
* Remove the `Component` suffix from components so their names have a 1-to-1 mapping to view directories. e.g. `app/views/users/user.rb` would map to `class Users::User < ViewComponent::Base`

A more complete example may look like:

```ruby
# app/controllers/users_controller.rb
class UsersController < ApplicationController
  def show
    user = User.find(params[:id])
    render :show, user: user
  end
end
```

```ruby
# app/views/users/show.rb
class Users::Show < ViewComponent::Base
  def initialize(user:)
    @user = user
  end
end
```

```erb
<%# app/views/users/show.html.erb %>
<div>
  <h1><%= @user.name %>'s page!</h1>
   <%= render "users/profile_content", user: user # will render the app/views/users/profile_content.rb component %>
</div>
```

part of https://github.com/github/view_component/issues/617

**Comment by MatheusRich** [comment]
Related: https://github.com/github/view_component/discussions/588

**Comment by BlakeWilliams** [comment]
I've spiked this out a bit here: https://github.com/BlakeWilliams/view_component-rails-test (PR's showing off other ideas or approaches are welcome)

Some take-aways:
* `_` prefixed file names for use as partials doesn't feel like the correct approach for components.
* Using strings, e.g. `render "users/profile, user: user"` instead of `render Users::Profile.new(user)` has some limitations, like requiring that component arguments are keyword arguments.
  * How would `render User::Profile.new(user)` translate to `render "user/profile`? The component would be required to change the initializer from `def initialize(user)` to `def initialize(user:)`.
* Implicit controller renders may be tricky or impractical.

I think the flexibility of view components (e.g. user defined initializers) may make it more challenging to make them work with Rails as seamlessly as we'd like, but I'll keep experimenting with it.

**Comment by choilive** [comment]
@BlakeWilliams I actually prefer the explicit object instantiation over using strings. Makes it easier to refactor and also makes it clearer that it is a "component" and not a partial. Would be OK with being able to use both ways though as a matter of user preference.

At the moment I feel like trying to unify view components with the existing rails approach to views seems like it would be a bit inconsistent . View_component encourages a more django-like "views and templates" approach that doesn't seem like it would mesh well.
Perhaps I just need to play around more with it and see how it would work in practice.

**Comment by boardfish** [comment]
#1412 brought this to my attention again. A couple of quick draft thoughts:

- It would be good to maintain compatibility with ActionView, for a few reasons:
  - It might take folks a good while to encapsulate their views
  - Some folks might not even want to encapsulate _all_ of their views for various reasons – developer friction, perhaps only using one or two helpers in a view, etc.
  Maybe the migration path looks something like this:
    1. User has an ActionView template
    2. User adds an associated class in the correct place, inheriting `ViewComponent::Base` and including something that makes sure helpers are still available
    3. User overrides helpers by adding methods on the component class, and replaces controller instance variables with variables on the component class
    4. User removes the include once they're through with the migration
- We may want to provide guidance for handing down things that might have been considered globally before from the controller to the view – things like the current user, tenant (for multitenanted apps), current time, etc.. These are the kinds of things that it helps to have available to us globally and might make a genuine use case for helpers, but it's also good to have visibility over them so that they can be mocked or stubbed in tests
- We could do with a codemod/RuboCop autocorrect for rewriting view specs as component specs

**Comment by Spone** [comment]
Relevant wizardry using Phlex: https://fly.io/ruby-dispatch/component-driven-development-on-rails-with-phlex/
I'm posting it here, since it might shed some light on the above points.

**Comment by bradgessler** [comment]
So awesome to see this come to life!

**Comment by joelhawksley** [comment]
@bradgessler @joeldrapper giving dhh's closing of the Rails PR, what do you think about writing a shared gem for implicit renderable rendering that both ViewComponent and Phlex could share?

**Comment by joeldrapper** [comment]
Yeah, that makes sense to me. Phlex-Rails could depend on that gem.

**Comment by joelhawksley** [comment]
@joeldrapper great. Want to pick a name? `rails-implicit-renderable-rendering`?

**Comment by reeganviljoen** [comment]
@joeldrapper @joelhawksley I really liked this idea, and just put together some code here, mind taking a look https://github.com/reeganviljoen/render_kit

**Comment by joelhawksley** [comment]
Closing as unplanned. Happy to see this land if desired some day.

**Comment by bradgessler** [comment]
@joelhawksley I have it working at https://github.com/rubymonolith/superview

Haven't released a gem yet, but if you give it a go from the main branch and it works how you think it should, I could cut a release.

**Comment by joelhawksley** [comment]
@bradgessler that's pretty cool! I think it looks good as is, but since we only do render explicity in the GitHub codebase I'll defer to others to test and validate ❤

**Comment by bradgessler** [comment]
In case any future web searchers stumble across this, I released Superview 1.0 today, which works with @joeldrapper's Phlex and @joelhawksley's ViewComponent libraries.

Article at https://terminalwire.com/articles/superview and repo at https://github.com/rubymonolith/superview.

I'll also note that it seems your first name must be Joel if you want to release a successful view component library.
